### PR TITLE
[master]get_from_assigner should return unreserved FCPs

### PR DESCRIPTION
**Pull request by SharpRazor**
Signed-off-by: SharpRazor <bjcb@cn.ibm.com>

### CODE REVIEW

**Comment by SharpRazor** [comment]
the difference of these 2 statement are following:
```
[root@113-cmp-cb databases]# sqlite3 sdk_fcp.sqlite
SQLite version 3.26.0 2018-12-01 12:34:55
Enter ".help" for usage hints.
sqlite> select * from fcp where assigner_id='CBM5400A' and (reserved <> 0 or connections <> 0);
1a11|CBM5400A|2|1|0|
1c14|CBM5400A|2|1|1|
1d18|CBM5400A|2|1|2|
sqlite> select * from fcp where assigner_id='CBM5400A';
1a10|CBM5400A|0|0|0|
1a11|CBM5400A|2|1|0|
1c14|CBM5400A|2|1|1|
1d18|CBM5400A|2|1|2|
sqlite>
```

**Comment by SharpRazor** [comment]
### without this patch:
```
[2021-03-19 01:35:23] [INFO] Got available fcp_list [('1a10', 'CBM5400A', 0, 0, 0, ''), ('1a11', 'CBM5400A', 1, 1, 0, ''), ('1c14', 'CBM5400A', 1, 1, 1, ''), ('1d18', 'CBM5400A', 1, 1, 2, '')] in Unreserve mode from CBM5400A.
```

### with this patch:
```
[2021-03-19 01:42:07] [INFO] Got available fcp_list [('1a11', 'CBM5400A', 1, 1, 0, ''), ('1c14', 'CBM5400A', 1, 1, 1, ''), ('1d18', 'CBM5400A', 1, 1, 2, '')] in Unreserve mode from CBM5400A.
```

File: zvmsdk/database.py [hunk 1]
```diff
@@ -458,6 +458,7 @@ def get_from_assigner(self, assigner_id):
         with get_fcp_conn() as conn:
 
             result = conn.execute("SELECT * FROM fcp WHERE assigner_id=? "
+                                  "and (reserved <> 0 or connections <> 0) "
```
Comment (dongyanyang) [inline zvmsdk/database.py:461]: then what's the difference between get_from_assigner and get_allocated_fcps_from_assigner?

**Review by dongyanyang — CHANGES_REQUESTED** [review]

File: zvmsdk/database.py (hunk 1 above)
Comment (jackydalong) [inline zvmsdk/database.py:461]: Suggest use capital format for all SQL keyword, ex:
image

Comment (SharpRazor) [reply zvmsdk/database.py:461]: good question, after change, they have no difference.

**Comment by jackydalong** [comment]
Suggest change below "LOG.debug" to "LOG.info", which makes debug from log clearer.
```python
#zvmsdk/volumeop.py
    def get_volume_connector(self, assigner_id, reserve):
        """Get connector information of the instance for attaching to volumes.
...
        LOG.debug('get_volume_connector returns %s for %s' %.       <<<<<
                  (connector, assigner_id))
```

File: zvmsdk/database.py (hunk 1 above)
Comment (SharpRazor) [reply zvmsdk/database.py]: I committed a new version, remove the condition that connections <> 0.

Comment (SharpRazor) [reply zvmsdk/database.py]: good advice, done.

**Comment by SharpRazor** [comment]
> @SharpRazor
> Suggest change below "LOG.debug" to "LOG.info", which makes debug from log clearer.
> 
> ```python
> #zvmsdk/volumeop.py
>     def get_volume_connector(self, assigner_id, reserve):
>         """Get connector information of the instance for attaching to volumes.
> ...
>         LOG.debug('get_volume_connector returns %s for %s' %.       <<<<<
>                   (connector, assigner_id))
> ```

OK, done.

**Review by jackydalong — APPROVED** [review]
LGTM

**Comment by jichenjc** [comment]
@dongyanyang please share your opinion, thanks

**Comment by jackydalong** [comment]
To make debug from log clearer, it better to print the fcp within related log.
Thus, highly suggest changing below LOG 
```python
zvmsdk/volumeop.py
    def _attach(self, fcp, assigner_id, target_wwpns, target_lun,
        LOG.info('Start to attach device to %s' % assigner_id)   <<<<<<<  1
            LOG.info('Attaching device to %s is done.' % assigner_id) <<<<<<< 2
        LOG.info('Attaching device to %s is done.' % assigner_id) <<<<<<< 3

    def _detach(self, fcp, assigner_id, target_wwpns, target_lun,
            LOG.info('Detaching device from %s is done.' % assigner_id)  <<<<< 4
        LOG.info('Detaching device from %s is done.' % assigner_id). <<<<<< 5 

```
to 
```python
    def _attach(self, fcp, assigner_id, target_wwpns, target_lun,
        LOG.info('Start to attach fcp device %s to %s' % (fcp, assigner_id)) 
            LOG.info('Attaching fcp device %s to %s is done.' % (fcp, assigner_id)) 
        LOG.info('Attaching fcp device %s to %s is done.' % (fcp, assigner_id)) 

    def _detach(self, fcp, assigner_id, target_wwpns, target_lun,
            LOG.info('Detaching fcp device %s from %s is done.' % (fcp, assigner_id))  
        LOG.info('Detaching fcp device %s from %s is done.' % (fcp, assigner_id)) 
```
@bjcb

File: zvmsdk/volumeop.py [hunk 2]
```diff
@@ -820,11 +821,7 @@ def _detach(self, fcp, assigner_id, target_wwpns, target_lun,
                                    need_restart)
                 raise exception.SDKBaseException(msg=errmsg)
 
-        # Unreserved fcp device after undedicate all FCP devices
```
Comment (SharpRazor) [inline zvmsdk/volumeop.py:823]: remove this because get_volume_connector will unreserve them.

**Comment by SharpRazor** [comment]
> To make debug from log clearer, it better to print the fcp within related log.
> Thus, highly suggest changing below LOG
> 
> ```python
> zvmsdk/volumeop.py
>     def _attach(self, fcp, assigner_id, target_wwpns, target_lun,
>         LOG.info('Start to attach device to %s' % assigner_id)   <<<<<<<  1
>             LOG.info('Attaching device to %s is done.' % assigner_id) <<<<<<< 2
>         LOG.info('Attaching device to %s is done.' % assigner_id) <<<<<<< 3
> 
>     def _detach(self, fcp, assigner_id, target_wwpns, target_lun,
>             LOG.info('Detaching device from %s is done.' % assigner_id)  <<<<< 4
>         LOG.info('Detaching device from %s is done.' % assigner_id). <<<<<< 5 
> ```
> 
> to
> 
> ```python
>     def _attach(self, fcp, assigner_id, target_wwpns, target_lun,
>         LOG.info('Start to attach fcp device %s to %s' % (fcp, assigner_id)) 
>             LOG.info('Attaching fcp device %s to %s is done.' % (fcp, assigner_id)) 
>         LOG.info('Attaching fcp device %s to %s is done.' % (fcp, assigner_id)) 
> 
>     def _detach(self, fcp, assigner_id, target_wwpns, target_lun,
>             LOG.info('Detaching fcp device %s from %s is done.' % (fcp, assigner_id))  
>         LOG.info('Detaching fcp device %s from %s is done.' % (fcp, assigner_id)) 
> ```
> 
> @bjcb

good advice, done.

**Comment by dongyanyang** [comment]
I need more time to check the logic, pls hold for a while

**Review by dongyanyang — APPROVED** [review]
LGTM

**Comment by jichenjc** [comment]
@SharpRazor you need solve the conflict.. thanks

**Pull request merged**